### PR TITLE
[GRPOTrainer]: Add SAPO Loss

### PR DESCRIPTION
**Pull request by pramodith**
# What does this PR do?
The Qwen team released a new paper [Soft Adaptive Policy Optimization](https://huggingface.co/papers/2511.20347) that introduces a new loss called SAPO that gets rid of hard clipping to token losses with a tempature controlled sigmoid gate.

> We propose Soft Adaptive Policy Optimization (SAPO), which
replaces hard clipping with a smooth, temperature-controlled gate that adaptively attenuates off-policy updates while preserving useful learning signals. 

The loss function is formulated as

image

with r and A being the importance ratios and Advantage scores

image

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [X] Did you make sure to update the documentation with your changes?
- [X] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: trl/trainer/grpo_config.py [hunk 1]
```diff
@@ -211,6 +219,10 @@ class GRPOConfig(TrainingArguments):
               clipped weights are then multiplied with the advantages and policy model's log probs. Individual token
               losses are aggregated by normalizing with the number of active tokens in the global accumulated batch.
               This method was introduced in the [MiniMax-M1 paper](https://huggingface.co/papers/2506.13585).
+            - `"sapo"`: Soft Adaptive Policy Optimization loss, as introduced in the [Soft Adaptive Policy Optimization
```
Comment (DaehanKim) [inline trl/trainer/grpo_config.py:220]: sapo already implies token-level ratio is being used for loss calculation
I think this can be addressed in loss_type and not duplicated here

File: trl/trainer/grpo_config.py [hunk 2]
```diff
@@ -627,6 +655,10 @@ class GRPOConfig(TrainingArguments):
             "Individual token losses are aggregated by normalizing with the number of active tokens in "
             "the global accumulated batch. This method was introduced in the "
             "[MiniMax-M1 paper](https://huggingface.co/papers/2506.13585)."
+            "'sapo': Soft Adaptive Policy Optimization loss, as introduced in the "
```
Comment (DaehanKim) [inline trl/trainer/grpo_config.py:663]: same as above comment, can be omitted here

**Review by DaehanKim — CHANGES_REQUESTED** [review]
Thank you for your contribution! Here's minor fix I want to suggest

File: tests/test_grpo_trainer.py [hunk 3]
```diff
@@ -167,11 +168,23 @@ def test_training(self, config_name):
             new_param = trainer.model.get_parameter(n)
             assert not torch.equal(param, new_param), f"Parameter {n} has not changed."
 
-    @pytest.mark.parametrize("loss_type", ["bnpo", "dr_grpo", "dapo", "cispo"])
+    def test_get_sapo_token_loss(self):
+        trainer = GRPOTrainer(
+            model="trl-internal-testing/tiny-Qwen2ForCausalLM-2.5",
+            reward_funcs=[accuracy_reward],
+        )
```
Comment (qgallouedec) [inline tests/test_grpo_trainer.py]: since it's a static method, I don't think you need to instantiate the trainer

File: trl/trainer/grpo_config.py [hunk 4]
```diff
@@ -172,6 +172,14 @@ class GRPOConfig(TrainingArguments):
...
             When used with `loss_type='cispo'`, this corresponds to the ε_max param specified in the [ScaleRL
             paper](https://arxiv.org/pdf/2510.13786) and the recommended value is `5.0`.
+        sapo_temperature_neg (`float`, *optional*):
+            Temperature for tokens with non-positive advantage scores used in the `sapo` loss function. If not
+            specified, it defaults to `1.05`. This parameter is introduced in the [Soft Adaptive Policy Optimization
+            paper](https://huggingface.co/papers/2511.20347).
+        sapo_temperature_pos (`float`, *optional*):
+            Temperature for tokens with positive advantage scores used in the `sapo` loss function. If not specified,
+            it defaults to `1.0`. This parameter is introduced in the [Soft Adaptive Policy Optimization
+            paper](https://huggingface.co/papers/2511.20347).
```
Comment (qgallouedec) [inline trl/trainer/grpo_config.py]:
```suggestion
        sapo_temperature_neg (`float`, *optional*, defaults to `1.05`):
            Temperature for tokens with non-positive advantage scores used in the `sapo` loss function. This parameter
            is introduced in the [Soft Adaptive Policy Optimization paper](https://huggingface.co/papers/2511.20347).
        sapo_temperature_pos (`float`, *optional*, defaults to `1.0`):
            Temperature for tokens with positive advantage scores used in the `sapo` loss function. This parameter is
            introduced in the [Soft Adaptive Policy Optimization paper](https://huggingface.co/papers/2511.20347).```

File: trl/trainer/grpo_config.py [hunk 5]
```diff
@@ -576,6 +588,22 @@ class GRPOConfig(TrainingArguments):
             "[ScaleRL paper]https://huggingface.co/papers/2510.13786) and the recommended value is `5.0`."
         },
     )
+    sapo_temperature_neg: float | None = field(
```
Comment (qgallouedec) [inline trl/trainer/grpo_config.py]:
```suggestion
    sapo_temperature_neg: float = field(
```

File: trl/trainer/grpo_config.py [hunk 6]
```diff
@@ -576,6 +588,22 @@ class GRPOConfig(TrainingArguments):
...
+        default=1.05,
+        metadata={
+            "help": "Temperature for tokens with non-positive advantage scores used in the `sapo` loss function. "
+            "If not specified, it defaults to `1.05`. This parameter is introduced in the "
+            "[Soft Adaptive Policy Optimization paper](https://huggingface.co/papers/2511.20347)."
+        },
+    )
+    sapo_temperature_pos: float | None = field(
```
Comment (qgallouedec) [inline trl/trainer/grpo_config.py]:
```suggestion
    sapo_temperature_pos: float = field(
```

File: trl/trainer/grpo_config.py [hunk 7]
```diff
@@ -576,6 +588,22 @@ class GRPOConfig(TrainingArguments):
             "[ScaleRL paper]https://huggingface.co/papers/2510.13786) and the recommended value is `5.0`."
         },
     )
+    sapo_temperature_neg: float | None = field(
+        default=1.05,
+        metadata={
+            "help": "Temperature for tokens with non-positive advantage scores used in the `sapo` loss function. "
+            "If not specified, it defaults to `1.05`. This parameter is introduced in the "
+            "[Soft Adaptive Policy Optimization paper](https://huggingface.co/papers/2511.20347)."
```
Comment (qgallouedec) [inline trl/trainer/grpo_config.py]:
```suggestion
            "help": "Temperature for tokens with non-positive advantage scores used in the `sapo` loss function. "
            "This parameter is introduced in the [Soft Adaptive Policy Optimization "
            "paper](https://huggingface.co/papers/2511.20347)."
```

File: trl/trainer/grpo_config.py [hunk 8]
```diff
@@ -576,6 +588,22 @@ class GRPOConfig(TrainingArguments):
...
+        },
+    )
+    sapo_temperature_pos: float | None = field(
+        default=1.0,
+        metadata={
+            "help": "Temperature for tokens with positive advantage scores used in the `sapo` loss function. "
+            "If not specified, it defaults to `1.0`. This parameter is introduced in the "
+            "[Soft Adaptive Policy Optimization paper](https://huggingface.co/papers/2511.20347)."
```
Comment (qgallouedec) [inline trl/trainer/grpo_config.py]:
```suggestion
            "help": "Temperature for tokens with positive advantage scores used in the `sapo` loss function. "
            "This parameter is introduced in the [Soft Adaptive Policy Optimization "
            "paper](https://huggingface.co/papers/2511.20347)."
```

File: docs/source/paper_index.md [hunk 9]
```diff
@@ -290,6 +290,25 @@ config = GRPOConfig(
...
+**📜 Paper**: https://huggingface.co/papers/2511.20347
+
+Soft Adaptive Policy Optimization (SAPO), replaces hard clipping with a smooth, temperature-controlled gate that adaptively attenuates off-policy updates while preserving useful learning signals. Compared with GSPO and GRPO, SAPO is both sequence-coherent and token-adaptive. Like GSPO, SAPO maintains sequence-level coherence, but its soft gating forms a continuous trust region that avoids the brittle hard clipping band used in GSPO.
+
+To reproduce the paper's setting, use this configuration:
+
+```python
+from trl import GRPOConfig
+
+config = GRPOConfig(
```
Comment (qgallouedec) [inline docs/source/paper_index.md]: for consistency

```suggestion
training_args = GRPOConfig(
```

File: trl/trainer/grpo_trainer.py [hunk 10]
```diff
@@ -1880,6 +1892,16 @@ def _compute_loss(self, model, inputs):
             per_token_loss1 = coef_1 * advantages
             per_token_loss2 = coef_2 * advantages
             per_token_loss = -torch.min(per_token_loss1, per_token_loss2)
+        elif self.loss_type == "sapo":
+            per_token_loss = torch.empty_like(coef_1)
+            pos_advantages_mask = advantages.repeat([1, coef_1.shape[1]]) > 0
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py]: can you use `positive` instead of `pos`

File: trl/trainer/grpo_trainer.py
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py:1958]: I'm wondering (not for this PR) if we should decouple the aggregation type and the loss type

File: tests/test_grpo_trainer.py [hunk 11]
```diff
@@ -167,11 +168,23 @@ def test_training(self, config_name):
...
+        # sigmoid(temp * (1-1)) * 4/temp = 0.5 * 4 = 2
+        expected_sapo_token_loss = torch.full_like(sample_token_importance_ratio, 2.0)
+        torch.testing.assert_close(sapo_token_loss, expected_sapo_token_loss)
+
+    @pytest.mark.parametrize("loss_type", ["bnpo", "dr_grpo", "dapo", "cispo", "sapo"])
     def test_training_loss_types(self, loss_type):
         dataset = load_dataset("trl-internal-testing/zen", "standard_prompt_only", split="train")
 
         training_args = GRPOConfig(
+            fp16=True,
```
Comment (qgallouedec) [inline tests/test_grpo_trainer.py]:
```suggestion
```

File: trl/trainer/grpo_trainer.py
Comment (pramodith) [reply trl/trainer/grpo_trainer.py:1958]: Yes! I find the current setup a bit annoying tbh.
